llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp

```diff
@@ -429,10 +431,14 @@ void LVELFReader::processOneAttribute(const DWARFDie &Die, LVOffset *OffsetPtr,
         CurrentHighPC = *Address;
       if (std::optional<uint64_t> Offset = FormValue.getAsUnsignedConstant())
         // High PC is an offset from LowPC.
-        CurrentHighPC = CurrentLowPC + *Offset;
+        // Don't add the WebAssembly offset if we have seen a DW_AT_low_pc.
+        CurrentHighPC =
+            (FoundLowPC ? CurrentLowPC - WasmOffset : CurrentLowPC) + *Offset;
```

Comment:
Why is this? Can you elaborate?

Reply:
When we detect the `DW_AT_low_pc`, the `CurrentLowPC` is updated to reflect the WebAssembly code section offset.
```
  case dwarf::DW_AT_low_pc:
      FoundLowPC = true;
      ...
      if (FoundLowPC) {
        ...
        // Consider the case of WebAssembly.
        CurrentLowPC += WasmOffset;
        ...
      }
      break;
```
When we detect the DW_AT_high_pc:
```
  case dwarf::DW_AT_high_pc:
        ...
        // Don't add the WebAssembly offset if we have seen a DW_AT_low_pc.
        CurrentHighPC =
            (FoundLowPC ? CurrentLowPC - WasmOffset : CurrentLowPC) + *Offset;
        ...
      // Consider the case of WebAssembly.
      CurrentHighPC += WasmOffset;
        ...
    break;
```
Basically if we have seen a `DW_AT_low_pc`, the `CurrentLowPC` is already increased by the `WasmOffset` value; so for the calculation of `CurrentHighPC` the value of `CurrentLowPC` must be the original value (no addendums).

Reply:
Please correct me where I'm mistaken.

Whether `CurrentLowPC` already contains the `WasmOffset` or not, `CurrentHighPC` should be higher than `CurrentLowPC`, right? The comment says
```cpp
// High PC is an offset from LowPC.
```
So if the high PC is 50 and the low PC is 40, the offset should be 10. (without `WasmOffset`)

Now add the `WasmOffset`. Suppose `WasmOffset` is 100, and `CurrentLowPC` is 140. And if we want to calculate `CurrentHighPC`, I think it still should be just
```cpp
CurrentHighPC = CurrentLowPC + *Offset;
```
So that `CurrentHighPC` is `CurrentLowPC ` (140) + `Offset` (10) = 150.

But this equation subtracts `WasmOffset` from it, so that
```cpp
CurrentHighPC = (CurrentLowPC - WasmOffset) + *Offset;
```
Here `CurrentHighPC` will be 140 - 100 + 10 = 50, whereas `CurrentLowPC` is 140, which is, `CurrentLowPC` contains `WasmOffset`, while `CurrentHighPC` does not.

I think this is strange. But given that your tool works, I must be mistaken somewhere. Can you point out where?

Reply:
Your analysis is correct until the line where `CurrentHighPC` is calculated based on `CurrentLowPC`.
Referring again to the code for `DW_AT_high_pc`:
```
  case dwarf::DW_AT_high_pc:
        ...
      if (std::optional<uint64_t> Offset = FormValue.getAsUnsignedConstant())
        // Don't add the WebAssembly offset if we have seen a DW_AT_low_pc.
        CurrentHighPC =
            (FoundLowPC ? CurrentLowPC - WasmOffset : CurrentLowPC) + *Offset;
```
At this point:
`CurrentHighPC` is 140 - 100 + 10 = **50**, whereas `CurrentLowPC` is **140**
```
        ...
      // Consider the case of WebAssembly.
      CurrentHighPC += WasmOffset;
```
At this point:
`CurrentHighPC` will be 50 + 100 = **150**, whereas `CurrentLowPC` is **140** and their offset is **10** (with WasmOffset)
```
        ...
    break;
```
I hope this clarify as how the values are calculated.

Reply:
Ah I see. Then, if we are going to add `WasmOffset` three lines later, why do we subtract it here?

Reply:
The `CurrentLowPC (DW_AT_low_pc)` always is encoded as an address.
The `CurrentHighPC (DW_AT_high_pc)` can be encoded as an address or as a constant.

When it is a constant, its value is an offset from the `CurrentLowPC`.
That is the only case where we use the formula:
`CurrentHighPC = (FoundLowPC ? CurrentLowPC - WasmOffset : CurrentLowPC) + *Offset
`
and we need to check if the `CurrentLowPC` has been already updated to include the `WasmOffset`.

If the `CurrentHighPC` is encoded as an address, we just add the `WasmOffset`.